***3. Programming Practice***

**1. Write a function that takes a list of integers and calculates their sum.**

In [6]:
def summation(lst):
    sum = 0
    for num in lst:
        sum = sum+num
    return sum

In [7]:
x = input("Enter the list of integers: ")
l = list(map(int,x.split()))
summation(l)

Enter the list of integers: 7 8 9 6 84 39


153

**2. Write a function that takes a list of integers and calculates the sum of all the odd values. Return the sum and the number of values encountered.** 

In [8]:
def oddsum(numbers):
  summ = 0
  for num in numbers:
    if(num%2!=0):
      summ = summ + num
  return(summ)

In [9]:
y = input("Enter the list of integers: " )
numbs = list(map(int,y.split()))
oddsum(numbs)

Enter the list of integers: 1 9 8 7 10 5


22

**3. Determine the max and min of three numbers.**

In [22]:
def maxmin(x,y,z):
    if (x>y):
        max = x
        min = y
    else:
        max = y
        min = x
    if z>x:
        max = z
    elif z<y:
        min = z
    return max, min

In [23]:
a = int(input("Enter first number: "))
b = int(input("Enter second number: "))
c = int(input("Enter third number: "))
maxmin(a,b,c)

Enter first number: 89
Enter second number: 86
Enter third number: 91


(91, 86)

**4. Determine if a given year is a leap year.**

In [14]:
def leapp(y):
    if (y%4!=0):
        return "Not a leap year"
    elif(y%100!=0):
        return "Leap year"
    elif(y%400==0):
        return "Leap year"
    else:
        return "Not a leap year"

In [18]:
def leapp(y):
    if (y%4!=0):
        return "Not a leap year"
    elif(y%100==0 and y%400==0 or y%100!=0):
        return "Leap year"
    else:
        return "Not a leap year"

In [29]:
def leapp(y):
    if(y%4==0 and y%100==0 and y%400==0 or y%4==0 and y%100!=0):
        return "Leap year"
    else:
        return "Not a leap year"

In [30]:
y = int(input("Enter the year: "))
leapp(y)

Enter the year: 1904


'Leap year'

**5. Implement factorial iteratively and recursively.**

In [5]:
def fact(n):
    f = 1
    for i in range(1,n+1):
        f = f*i
    return f

In [6]:
x = int(input("Enter the number for factorial: "))
fact(x)

120

**6. Calculate the nth Fibonacci number iteratively and recursively.**

In [37]:
def fibo(n):
    fibo = [0,1]
    for i in range(2,n):
      a = fibo[i-2] + fibo[i-1]
      fibo.append(a)
    return fibo[-1]

In [38]:
x = int(input("Enter the nth Fibonacci number: "))
fibo(x)

Enter the nth Fibonacci number: 5


3

**7.Write a function that checks if a given word is a palindrome.**

In [5]:
def palindrome(word):
    i = 0
    j = len(word) - 1
    while(i<j):
        if (word[i]!= word[j]):
            return "Not a Palidnrome word"
        i = i + 1
        j = j - 1
    return "Palidnrome word"

In [6]:
s = str(input())
palindrome(s)

civic


'Palidnrome word'

**8. Generate a list of random integers. Take input 
 from the user (an integer value) and search the list for 
. Return the index of 
 in the list, if it exists, or None otherwise.**

In [47]:
import random
def rndm(rndnums, k, integer_number):
    integer_number = [random.randint(1, 100) for i in range(rndnums)]
    try:
        y = integer_number.index(k)
        print(y)
    except ValueError:
        print(None)

In [48]:
numss = int(input("Required numbers in the list: "))
m = int(input("Find the index of value: "))
integerss = [random.randint(1, 100) for i in range(numss)]
rndm(numss, m, integerss)

Required numbers in the list: 100
Find the index of value: 10
61


**9th. Calculate $e^x$, sin(x) and cos(x)recursively.**

In [15]:
import math
def trigo():
    x = int(input())
    degree = math.radians(x)
    c = 0
    while(c==0):
        ans  = round(math.exp(x) * 10**13,2)
        sinn = round(math.sin(degree),2)
        coss = round(math.cos(degree),2)
        print(ans)
        print(sinn)
        print(coss)
        c=c+1
    
trigo()

30
1.0686474581524462e+26
0.5
0.87


**1. e^x iteration**

In [3]:
x = 7.393922
round(x,4)

7.3939

In [8]:
ans = pow(2,3)
print(ans)

8


In [13]:
t = (pow(0,1)) / fact(0)
print(t)

0.0


In [12]:
fact(1)

1

In [23]:
def eul(x):
    t = 1
    i = 1
    val = 1
    while(round(t,10)!=0):
        t = (pow(x,i)) / fact(i)
        val = val + t
        i = i + 1
    return val


In [26]:
eul(3)

20.085536923183504

In [28]:
X = 41/24
X

1.7083333333333333

**10. Check if a Number is an Armstrong Number. An Armstrong number is a number that is the sum of its own digits each raised to the power of the number of digits. For example, 153 = 1^3 + 5^3 + 3^3.**

In [1]:
import math
def arm(x):
    y = len(x)
    summ = 0
    for i in x:
        cub = math.pow(int(i),y)
        summ = summ + cub
    if int(x) == summ:
        return 'Armstrong number'
    else:
        return 'Not an Armstrong number'

In [3]:
u = input("Enter a number to check: ")
arm(u)


'Not an Armstrong number'

In [1]:
def lnnum(n):
    c=0
    while(n>0):
        n = n//10
        c = c + 1
    return c

In [6]:
lnnum(15378)

5

In [18]:
# def arms(z,y):
#     arm = 0
#     while(z>0):
#         rem = z%10
#         arm = rem**y + arm
#         z = z//10
#     if(arm==z):
#         return True
#     else:
#         return False


In [19]:
# a = int(input("Enter a number"))
# x = lnnum(a)
# arms(a,x)

False

**11. Take a 3 digit number from the user, with different first and last digits. Reverse it and subtract the smaller from the larger. Call this d. Reverse d and add it to d. Print the result.**

In [1]:
def revv(num):
    if(len(num)==3 and num[0]!=num[2]):
        rev = num[-1::-1]
        print(rev)
        if(int(num)>int(rev)):
            d = int(num) - int(rev)
            d1 = str(d)
            revd = d1[-1::-1]
            revd1 = int(revd)
            ans = revd1 + d
            print(ans)
        else:
            d = int(rev) - int(num)
            d1 = str(d)
            revd = d1[-1::-1]
            revd1 = int(revd)
            ans = revd1 + d
            print(ans)
    else:
        print("Invalid Input")

In [2]:
z = input("Enter the three digit number: ")
revv(z)

Enter the three digit number: 153
351
1089


* Lenght of an integer

In [2]:
def lnnum(n):
    c=0
    while(n>0):
        n = n//10
        c = c + 1
    return c

In [3]:
lnnum(26948)

5

In [33]:
import math
def reversee(z):
    rev = 0
    while(z>0):
        rem = z%10
        rev = rem*pow(10,lnnum(z)-1) + rev
        z = z//10
    return rev


In [35]:
reversee(990)

99

In [30]:
def mag(x):
    if(x>reversee(x)):
        d = x-reversee(x)
        if(lnnum(d)==2):
            d = d*10
    else:
        d = reversee(x)-x
        if(lnnum(d)==2):
            d = d*10
    summ = d + reversee(d)
    return summ

In [31]:
mag(351)

1089

In [32]:
m = input("Enter three digit num")
if(len(m)==3 and m[0]!=m[2]):
    print(mag(int(m)))
else:
    print("Invalid Input")

1089


**ITERTATIONS**

**1. Count Multiples: Given a positive integer n, count how many numbers from 1 up to n are divisible by 3.**

In [6]:
def multi(n): # Define function 'multi' that takes in one paramter 'n'
    c = 0 # Initializing a variable 'c' to store the count of numbers divisible by 3
    for i in range(1,n+1): # Loop through numbers 1 upto 'n' (includes n)
        if(i%3==0): # Check if the number 'i' is divisble by 3
            c = c+1 # If YES then increment the count by 1
    return c # return the final count of numbers divisble by 3

In [7]:
multi(10) # calling the function with 10 as the input parameter

3

In [11]:
assert(multi(20)==8)

AssertionError: 

**2. Running Total: You are given a list of numbers representing daily expenses. Starting from zero, add them one by one and report the final total.**



In [1]:
def runtot(lst): # Define function 'runtot' that takes in a list of integers
    total = 0 # Iniialize variable 'total' to store the total
    for i in lst: # Loop that iterates through all the numbers in the list
        total = total + i # Add to previous total
    return total # Return final total 

In [2]:
import random # library random  
explist = [random.randint(1,10) for i in range(10)] # Use list comprehension to store get 10 random numbers that lie between 1-100
print(explist) 
runtot(explist) # Calling the function with list of randome integers


[1, 1, 4, 5, 10, 3, 1, 6, 8, 5]


44

In [3]:
assert(runtot(explist)==sum(explist))

**3. First Negative: Given a list of integers, scan from left to right and report the index of the first negative number. If there is no negative number, report -1.**

In [20]:
def frstneg(listt):
    for i in range(len(listt)): # Iterate through all  number in the list
        if(listt[i]<0): # Check if the current number is negative
            return i
    return -1 # If no negative number then return 1

In [21]:
import random # library random  
randlist = [random.randint(-2,100) for i in range (50)] # Use list comprehension to store get 10 random numbers that lie between 1-100
print(randlist)
frstneg(randlist)

[6, 25, 78, 16, 6, 6, 44, 65, 7, 11, 30, 79, 92, 33, 90, 98, 56, 34, 70, 55, 76, 73, 41, 40, 1, 31, 37, 22, 40, -2, 70, 18, 58, 43, 93, 32, 89, 54, 59, 77, 10, 42, 5, 98, 12, 86, 47, 16, 67, 51]


29

**4. Password Attempts: You are given a list of password attempts in the order they were tried. The correct password is "python123". Determine whether the correct password ever appears in the list.**

In [1]:
def pw(passlist):
    for i in passlist: # Iterate through list of passwords
        if (i=="python123"): # Check if the list has correct passoword
            return "correct password"
    return "incorrect password"

In [2]:
pwlist = ["python", "py123", "python121", "python123"]
pw(pwlist)

'correct password'

In [3]:
assert(pw(pwlist)=="correct password")

**5. Temperature Check: You are given a list of daily temperatures. Determine whether all of them are within the safe range of 18 to 30 degrees inclusive.**

In [26]:
def tempcheck(temps):
        check = all(18<=i<=30 for i in temps) # check stores output of all function which includes iteration through the list
        return check

In [27]:
import random
temperatures = [random.randint(10,40) for i in range(10)] # Use list comprehension to store get 10 random temperatures that lie between 10-40
print(temperatures)
tempcheck(temperatures)

[28, 33, 21, 17, 20, 30, 23, 36, 20, 13]


False

**6. Longest Streak: Given a list of booleans representing whether a task was completed each day, determine the length of the longest consecutive streak of True values.**

In [59]:
def streak(listt):
    c = 0 # Counting the streak of "True"
    maxx = 0 # Storing the longest streak of "True"
    for i in listt:
        if(i == True):
            c = c + 1
            if(c > maxx):
                maxx = c
        else:
            c = 0
    return maxx
    

In [60]:
listt = [False, True, True, False, True, True, True, False, True, True, True, True, True, False, True]
streak(listt)

5

**8. Balance After Transactions: You start with a bank balance of zero. You are given a list of transactions (positive numbers are deposits, negative numbers are withdrawals). Process them in order and determine whether the balance ever goes negative at any point.**



In [83]:
def bal(trans):
    balance = 0
    for i in trans:
        balance = balance + i
        if(balance < 0):
            print(balance)
            return True

In [87]:
import random
Transs = [random.randint(-50, 50) for i in range (20)]
print(Transs)
bal(Transs)



[33, -3, 15, -26, -38, -13, -47, 42, 42, -11, 17, 15, 50, 10, 23, -2, -9, -38, -4, 32]
-19


True

**8th. First Repeated Item: Given a list of integers, find the first value that appears more than once when scanning from left to right. If no value repeats, report that fact.**

In [22]:
def repeated(items):
    for i in range(len(items)-1):
        if (items[i] == items[i+1]):
            return items[i]
    return "No repeated items"

In [23]:
import random
intgrs = [random.randint(-100, 100) for i in range (20)]
intgrs.sort()
print(intgrs)
repeated(intgrs)

[-79, -77, -57, -55, -38, -20, -6, -2, 4, 4, 9, 15, 18, 35, 51, 65, 72, 72, 79, 100]


4

**9th. Filtered Average: Given a list of numbers, compute the average of only those values that lie between 10 and 100 inclusive. If no such values exist, return None.**

In [35]:
def filavg(ints):
    filints = [i for i in ints if(10<=i<=100)]
    print(filints)
    return round(sum(filints) / len(filints),2)

In [37]:
import random
avgintgrs = [random.randint(-10, 200) for i in range (10)]
print(avgintgrs)
filavg(avgintgrs)


[99, 43, 90, 127, 200, 9, 22, 146, 10, 142]
[99, 43, 90, 22, 10]


52.8

**10. Stop on Threshold: You are given a list of numbers and a threshold value. Starting from the beginning of the list, keep adding numbers until the running total becomes strictly greater than the threshold. Report how many numbers were needed.**

In [41]:
def threshold (nums, thresh):
    total = 0
    c = 0
    for i in nums:
        total = total + i
        c = c + 1
        if (total > thresh):
            return c


In [43]:
nums = [10, 20, 24, 30, 59]
thresh = 80
threshold(nums, thresh)

4

**11. Prime Counter: Given a positive integer n, count how many prime numbers exist between 2 and n inclusive.**

In [57]:
def prime(n):
    if(n<=1):
        return False
    for i in range(2,n):
        if(n%i==0):
            return False
    return True

In [58]:
prime(7)

True

In [59]:
def pc(nums):
    c = 0
    for i in range(2, nums + 1):
        if(prime(i)):
            c = c + 1
    return c

In [60]:
pc(10)

4

**15. Sliding Window Maximum Sum: Given a list of integers and a window size k, determine the maximum sum of any k consecutive elements in the list.**

In [10]:
elements = [10, 40, 30, 20, 25, 50, 10, 5, 25]
k = 3
j = 0
maxi_sum = 0

for j in range(len(elements) - k + 1):
    summ = 0

    for i in range(j, k + j):
        summ = summ + elements[i]

    if(summ > maxi_sum):
        maxi_sum = summ

    print(summ)  
print(maxi_sum)    
    
     



80
90
75
95
85
65
40
95
